domain/network/state/linklayer_merge.go

```diff
@@ -0,0 +1,717 @@
+SELECT &address.*
+FROM ip_address AS ip
+LEFT JOIN provider_ip_address AS pip ON ip.uuid = pip.address_uuid
+WHERE ip.device_uuid = ($device.uuid)`, address{}, device{})
+	if err != nil {
+		return nil, errors.Capture(err)
+	}
+
+	var devices []device
+	if err := tx.Query(ctx, getDevicesStmt,
+		netNode{UUID: netNodeUUID}).GetAll(
+		&devices); err != nil && !errors.Is(err, sqlair.ErrNoRows) {
+		return nil, errors.Errorf(
+			"getting all link layer devices from net notd %q: %w",
+			netNodeUUID, err)
+	}
+
+	var result []mergeLinkLayerDevice
+	for _, device := range devices {
+		var addresses []address
+		if err := tx.Query(ctx, getAddressesStmt,
+			device).GetAll(&addresses); err != nil &&
+			!errors.Is(err, sqlair.ErrNoRows) {
+			return nil, errors.Errorf(
+				"getting all addresses for link layer device %q: %w",
+				device.UUID, err)
+		}
+		deviceType, err := decodeDeviceType(device.TypeID)
+		if err != nil {
+			return nil, errors.Errorf(
+				"decoding device type %d: %w", device.TypeID, err)
+		}
+		result = append(result, mergeLinkLayerDevice{
+			UUID:       device.UUID,
+			Name:       device.Name,
+			MACAddress: device.MACAddress,
+			ProviderID: device.ProviderID,
+			Type:       deviceType,
+			Addresses: transform.Slice(addresses,
+				func(a address) mergeAddress {
+					return mergeAddress{
+						UUID:       a.UUID,
+						Value:      a.Value,
+						ProviderID: a.ProviderID,
+					}
+				}),
+		})
+	}
+
+	return result, nil
+}
+
+// getMachineNetNodeUUID retrieves the NetNodeUUID associated with a given
+// Machine UUID from the database using a prepared SQL statement.
+// Returns the NetNodeUUID or an error if the operation fails.
+func (st *State) getMachineNetNodeUUID(
+	ctx context.Context, tx *sqlair.TX,
+	machineUUID string,
+) (string, error) {
+	type node struct {
+		MachineUUID string `db:"machine_uuid"`
+		NetNodeUUID string `db:"net_node_uuid"`
+	}
+
+	machine := node{MachineUUID: machineUUID}
+
+	stmt, err := st.Prepare(
+		`
+SELECT &node.net_node_uuid 
+FROM   machine 
+WHERE  uuid = $node.machine_uuid
+`, machine,
+	)
+	if err != nil {
+		return "", errors.Capture(err)
+	}
+
+	return machine.NetNodeUUID, errors.Capture(
+		tx.Query(
+			ctx, stmt, machine,
+		).Get(&machine),
+	)
+}
+
+// matchByName matches the incoming devices by name.
+//
+// It returns a map of the devices by name. If there are duplicate names,
+// the first one is used.
+func (st *State) matchByName(
+	ctx context.Context, normalized []mergeLinkLayerDevice,
+) map[string]mergeLinkLayerDevice {
+	result := make(map[string]mergeLinkLayerDevice, len(normalized))
+	for _, netInterface := range normalized {
+		if _, found := result[netInterface.Name]; found {
+			st.logger.Debugf(
+				ctx, "duplicate name %q in incoming network"+
+					" interfaces", netInterface.Name,
+			)
+			continue
+		}
+		result[netInterface.Name] = netInterface
+	}
+	return result
+}
+
+// normaliseIncoming is intended to accommodate providers such as EC2
```

Comment:
This comment does not make the intended purpose of this function very clear, even though it explains the specifics well, maybe it could start with something saying that it is intended to match incoming devices to existing devices where possible.